### PR TITLE
fix: bug fix when group by a last joined table

### LINKED ISSUE #911: After lastjoin, group by gets empty set
Issue tracker is **ONLY** used for reporting bugs. New features should be discussed on our discussion
After lastjoin, group by gets empty set

## Expected Behavior
        - [ "aa",26 ]
        - [ "cc",151 ]

## Current Behavior
Empty set

## case
```yaml
  - id: 21
    desc: lastjoin后group by
    inputs:
      - columns: [ "c1 string","c2 int","c3 bigint","c4 timestamp" ]
        indexs: [ "index1:c1:c4" ]
        rows:
          - [ "aa",2,3,1590738989000 ]
          - [ "aa",21,31,1590738990000 ]
          - [ "cc",41,51,1590738991000 ]
      - columns: [ "c1 string","c2 int","c3 bigint","c4 timestamp" ]
        indexs: [ "index1:c1:c3" ]
        rows:
          - [ "aa",2,13,1590738989000 ]
          - [ "bb",21,131,1590738990000 ]
          - [ "cc",41,151,1590738992000 ]
    sql: select {0}.c1,sum({1}.c3) as v1 from {0} last join {1} ORDER BY {1}.c3 on {0}.c1={1}.c1 group by {0}.c1;
    expect:
      order: c1
      columns: [ "c1 string","v1 bigint"]
      rows:
        - [ "aa",26 ]
        - [ "cc",151 ]
```

### PR DESCRIPTION
**Pull request by jingchen2222**
* **Please check if the PR fulfills these requirements**

- The commit message follows our guidelines
- Tests for the changes have been added (for bug fixes / features)
- Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix
* **What is the current behavior?** (You can also link to an open issue here)
Resolve #911 

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: hybridse/src/vm/runner.cc [hunk 1]
```diff
@@ -1967,9 +1967,9 @@ bool JoinGenerator::PartitionJoin(std::shared_ptr<PartitionHandler> left,
         LOG(WARNING) << "fail to run last join: left input empty";
         return false;
     }
-    if (!left_key_gen_.Valid()) {
+    if (!index_key_gen_.Valid() && !left_key_gen_.Valid()) {
         LOG(WARNING) << "can't join right partition table when join "
-                        "left_key_gen_ is invalid";
+                        "left_key_gen_ and index_key_gen_ are invalid";
```
Comment (aceforeverd) [inline hybridse/src/vm/runner.cc:1971]: neither left_key_gen_ or index_key_gen_ is valid

File: hybridse/src/vm/runner.cc [hunk 2]
```diff
@@ -1984,10 +1984,15 @@ bool JoinGenerator::PartitionJoin(std::shared_ptr<PartitionHandler> left,
...
-                index_key_gen_.Valid() ? index_key_gen_.Gen(left_row, parameter) + "|" +
-                                             left_key_gen_.Gen(left_row, parameter)
-                                       : left_key_gen_.Gen(left_row, parameter);
+
+            std::string key_str = "";
+            if (index_key_gen_.Valid()) {
+                key_str = index_key_gen_.Gen(left_row, parameter);
+            }
+            if (left_key_gen_.Valid()) {
+                key_str = key_str.empty() ? left_key_gen_.Gen(left_row, parameter) : key_str.append("|").append
```
Comment (aceforeverd) [inline hybridse/src/vm/runner.cc:1992]: use `absl::StrAppend` ?

File: hybridse/src/vm/runner.cc [hunk 3]
```diff
@@ -1967,9 +1967,9 @@ bool JoinGenerator::PartitionJoin(std::shared_ptr<PartitionHandler> left,
         LOG(WARNING) << "fail to run last join: left input empty";
         return false;
     }
-    if (!left_key_gen_.Valid()) {
+    if (!index_key_gen_.Valid() && !left_key_gen_.Valid()) {
```
Comment (aceforeverd) [inline hybridse/src/vm/runner.cc:1969]: Cool this function is not covered by test previously

File: hybridse/src/vm/runner.cc [hunk 4]
```diff
@@ -1984,10 +1984,15 @@ bool JoinGenerator::PartitionJoin(std::shared_ptr<PartitionHandler> left,
...
-            const std::string& key_str =
-                index_key_gen_.Valid() ? index_key_gen_.Gen(left_row, parameter) + "|" +
-                                             left_key_gen_.Gen(left_row, parameter)
-                                       : left_key_gen_.Gen(left_row, parameter);
+
+            std::string key_str = "";
+            if (index_key_gen_.Valid()) {
+                key_str = index_key_gen_.Gen(left_row, parameter);
+            }
+            if (left_key_gen_.Valid()) {
```
Comment (aceforeverd) [inline hybridse/src/vm/runner.cc:1991]: we might cover this lines as well with additional equal condition

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by dl239 — APPROVED** [review]

**Review by aceforeverd — APPROVED** [review]

**Comment by lumianph** [comment]
[bugfix tracking](https://trello.com/c/wEQ9HW0k/25-bugfix-tracking)

**Pull request merged**